GitHub Issue NOAA-EMC/GSI#341. Update the use of MW radiances for GFS v16.3 implementation

emilyhcliu opened the pull request:
**The following changes will be added to GSI Master for v16.3 Implementation:**

**Source code update:**
- [x] Switching from assimilating antenna temperature (SDR) to brightness temperature (TDR) for AMSU-A, ATMS, and MHS
- [x] The use precipitation-affected radiances for ATMS and AMSU-A (extened all-sky framework)
- [x] The use of GMI radiances under all-sky conditions

**Fix files update:**
- [x] Add global_anavinfo_allhydro_l127.txt --- add precipitation components
- [x] Update cloudy_radiance_info.txt --- add GMI obs error assignment and QC parameters
- [x] Update global_satinfo.txt --- modify channel use flag and error bounds for GMI
- [x] Update global_scaninfo.txt --- add GMI

**Script update:**
- [x] exglobal_atmos_analysis.sh

**Working Branch:**  **[feature/v16x_allsky](https://github.com/emilyhcliu/GSI/tree/feature/v16x_allsky)**

- The code update in the branch will not change the GSI result from master with the default namelist and default resource files (All regression tests passed) 
- Set the logical ta2tb to true in the namelist to activate the use of brightness temperature for ATMS, AMSU-A and MHS
- Use global_anavinfo_allhydro.l127.txt to enable the assimilation of precipitation-affected radiances from ATMS and AMSUA
- Modify the channel use flags for GMI in global_satinfo.txt to turn on the assimilation of GMI (see below for channel selection)

**List of changes for Ta2Tb work:**
- antcorr_application.f90
- crtm_interface.f90
- read_atms.f90
- read_bufrtovs.f90
- obsmod.F90
- gsimod.F90

Set ta2tb = .true. will trigger the use of Tb for AMSU-A, ATMS, and MHS
Default ta2tb is set to false.   This ensures that the code update will not change GSI result from the master

**List of changes for assimilating precipitation-affected ATMS and AMSU-A:** 
- cloud_efr_mod.f90
- cplr_gfs_ensmod.f90
- crtm_interface.f90
- general_read_gfsatm.f90
- netcdfgfs_io.f90
- prt_guess.f90
- qcmod.f90
- read_atms.f90
- read_bufrtovs.f90
- write_incr.f90
- setuprad.f90

These code changes will not affect GSI result from master unless the default global_anavinfo.l127.txt is replaced with global_anavinfo_alhydro.l127.txt

The current GSI master (and operational GDAS) is using CRTM2.3.0 for radiance assimilation
The use of precipitation-affected  MW data should combined with the use of CRTM 2.4.0 for the following two reasons:

1. The surface Jacobian under scattering condition from CRTM2.3.0 was not implemented correctly and CRTM 2.4.0 fixed the issue.  (Surface Jacobian is used in QC for radiance data)
2. CRTM2.4.0 has the cloud optical table generated based on GFDL physics (consistent with FV3 model physics)

Note: If we decide not to include precipitation-affected MW radiance in our next implementation, then we should not use CRTM2.4.0 (should stay with CRTM 2.3.0) due to the following two reasons:

1. CRTM 2.4.0 requires cloud fraction profile in the input for cloudy simulation.  
2. GFDL cloud fraction calculation requires all hydrometeors which is not available in our operational run (only cloud water and cloud water are available)

**List of changes for the assimilation of GMI under all-sky conditions:**

- read_gmi.f90
- crtm_interface.f90
- setuprad.f90

These code update  will not change GSI result from the master since GMI channel use flags are set to -1 (monitoring)
To activate GMI for assimilation, modifiy the use flags as the following:
```
 gmi_gpm                 1  -1    2.700   25.000   30.000   40.000    0.000      1      1     -1
 gmi_gpm                 2  -1    3.700   40.000   30.000   40.000    0.000      1      1     -1
 gmi_gpm                 3  -1    3.500   40.000   15.000   40.000    0.000      1      1     -1
 gmi_gpm                 4  -1    4.500   55.000   30.000   40.000    0.000      1      1     -1
 gmi_gpm                 5   1    4.000   35.000   15.000   40.000    0.000      1      1     -1
 gmi_gpm                 6   1    3.800   25.000   30.000   40.000    0.000      1      1     -1
 gmi_gpm                 7   1  300.000  500.000    5.000   40.000    0.000      1      1     -1
 gmi_gpm                 8  -1    5.000   50.000   15.000   40.000    0.000      1      1     -1
 gmi_gpm                 9  -1   11.500   50.000   20.000   40.000    0.000      1      1     -1
 gmi_gpm                10   1    5.000   50.000   20.000   40.000    0.000      1      1     -1
 gmi_gpm                11  -1    5.000   50.000   20.000   40.000    0.000      1      1     -1
 gmi_gpm                12   1    2.500   30.000   10.000   40.000    0.000      1      1     -1
 gmi_gpm                13   1    3.000   40.000   10.000   40.000    0.000      1      1     -1
```

**List of changes for exglobal_atmos_analysis.sh:**

1. add link to AMSU-A MetOp-A version 2 coefficient file
2. add logic (ta2tb) to turn on the use of brightness temperature for ATMS, AMSU-A and MHS for assimilaton
3. add logic (cao_check) to turn on CAO (cold air outbreak check) for quality control
4. set thinning box to 145 km for GMI (original is 100 km)

emilyhcliu committed:
GitHub Issue NOAA-EMC/GSI#341

This update includes the following:
(1) Add the capability of assimilating precipitation-affected radiances from AMSU-A and ATMS
(2) Enable the all-sky radiance assimilation for GMI
(3) Add the switch to use brightness temperature (SDR) instead of antenna temperature (TDR) for AMSU-A, ATMS and MHS

emilyhcliu commented:
@MichaelLueken-NOAA I suggest the following people for our internal review of this PR:
@ScottSieron-NOAA 
@ADCollard

MichaelLueken commented on src/gsi/cplr_gfs_ensmod.f90:
```diff
@@ -1015,7 +1057,8 @@ subroutine parallel_read_gfsnc_state_(en_full,m_cvars2d,m_cvars3d,nlon,nlat,nsig
    deallocate(rwork3d1)
 
-   if (k3u==0.or.k3v==0.or.k3t==0.or.k3q==0.or.k3cw==0.or.k3oz==0) &
+!  if (k3u==0.or.k3v==0.or.k3t==0.or.k3q==0.or.k3cw==0.or.k3oz==0) &  !orig
```
If commented out line 1060 won't be used in the future, please remove it.

MichaelLueken commented on src/gsi/netcdfgfs_io.f90:
```diff
@@ -116,8 +116,10 @@ subroutine read_
+    use constants, only: qcmin
     use gridmod, only: sp_a,grd_a,lat2,lon2,nsig
-    use guess_grids, only: ifilesig,nfldsig
+    ! use guess_grids, only: ifilesig,nfldsig
```
If commented out line 121 won't be used in the future, please remove it.

ADCollard commented on src/gsi/qcmod.f90:
```diff
@@ -3132,12 +3130,34 @@ subroutine qc_amsua(nchanl,is,ndat,nsig,npred,sea,land,ice,snow,mixed,luse,   &
+                !    enddo
+                !    errf(1:ich544)=zero
+                !    varinv(1:ich544)=zero
+                !    do i=1,ich544
```
Remove commented code

emilyhcliu replied:
@ADCollard These lines are for future testing. So, they are commented out for now.

ADCollard commented on src/gsi/read_bufrtovs.f90:
```diff
@@ -506,35 +521,77 @@ subroutine read_bufrtovs(mype,val_tovs,ithin,isfcalc,&
      call openbf(lnbufr,'IN',lnbufr)
 
-     if(llll >= 2 .and. (amsua .or. amsub .or. mhs))then
+     ! sieron 02/21 support multiple spc coefficient files for any given sensor
```
no need to comment about re-doing indentation!

ADCollard commented on src/gsi/read_bufrtovs.f90:
```diff
@@ -633,6 +690,18 @@ subroutine read_bufrtovs(mype,val_tovs,ithin,isfcalc,&
            call map2tgrid(dlat_earth,dlon_earth,dist1,crit1,itx,ithin,itt,iuse,sis,it_mesh=it_mesh)
            if(.not. iuse)cycle read_loop
 
+           ! sieron
```
Remove Sieron comment

emilyhcliu replied:
Done!

ADCollard commented on src/gsi/read_bufrtovs.f90:
```diff
@@ -476,7 +490,8 @@ subroutine read_bufrtovs(mype,val_tovs,ithin,isfcalc,&
   nreal = maxinfo + nstinfo
   nele  = nreal   + nchanl
-  hdr1b ='SAID FOVN YEAR MNTH DAYS HOUR MINU SECO CLAT CLON CLATH CLONH HOLS'
+  ! sieron add SACV
```
remove sieron

ADCollard commented on src/gsi/read_bufrtovs.f90:
```diff
@@ -234,6 +242,12 @@ subroutine read_bufrtovs(mype,val_tovs,ithin,isfcalc,&
   logical :: critical_channels_missing,quiet
   logical :: print_verbose
 
+  ! sieron
```
remove sieron

ADCollard commented on src/gsi/read_bufrtovs.f90:
```diff
@@ -190,6 +196,8 @@ subroutine read_bufrtovs(mype,val_tovs,ithin,isfcalc,&
 
   integer(i_kind) ireadsb,ireadmg,irec,next,nrec_startx
   integer(i_kind) i,j,k,ifov,ntest,llll
+  ! sieron add this variable
```
remove comment

ADCollard commented on src/gsi/read_bufrtovs.f90:
```diff
@@ -174,7 +179,8 @@ subroutine read_bufrtovs(mype,val_tovs,ithin,isfcalc,&
   character(8),parameter:: fov_flag="crosstrk"
-  integer(i_kind),parameter:: n1bhdr=13
+  !sieron change from 13 to 14 for sacv
+  integer(i_kind),parameter:: n1bhdr=14
```
remove comment

ADCollard commented on src/gsi/read_bufrtovs.f90:
```diff
@@ -141,9 +143,12 @@ subroutine read_bufrtovs(mype,val_tovs,ithin,isfcalc,&
   use calc_fov_crosstrk, only : instrument_init, fov_cleanup, fov_check
   use gsi_4dvar, only: l4dvar,l4densvar,iwinbgn,winlen
-  use antcorr_application, only: remove_antcorr
+  ! sieron
```
remove comment

ADCollard commented on src/gsi/read_bufrtovs.f90:
```diff
@@ -141,9 +143,12 @@ subroutine read_bufrtovs(mype,val_tovs,ithin,isfcalc,&
       crtm_kind => fp, &
       MAX_SENSOR_ZENITH_ANGLE
   use crtm_spccoeff, only: sc,crtm_spccoeff_load,crtm_spccoeff_destroy
```
remove comment

ADCollard commented on src/gsi/read_bufrtovs.f90:
```diff
@@ -676,20 +745,65 @@ subroutine read_bufrtovs(mype,val_tovs,ithin,isfcalc,&
-                 end do
+!          sieron: changed from accepting TMBR and converting TMBRST to
+!          antenna temperature (remove_antcorr), to converting TMBR to
+!          brightness temperature (add_antcorr) and accepting TMBRST
```
Describe what the code does, not the changes

ADCollard commented on src/gsi/read_bufrtovs.f90:
```diff
@@ -676,20 +745,65 @@ subroutine read_bufrtovs(mype,val_tovs,ithin,isfcalc,&
+                    data1b8x=data1b8
+                    data1b4=data1b8
+                    call remove_antcorr(accoeff_sets(sacv),ifov,data1b4)
+                    !call apply_antcorr(accoeff_sets(spc_coeff_versions),ifov,data1b4)
```
remove comment

ADCollard commented on src/gsi/read_gmi.f90:
```diff
@@ -131,7 +131,12 @@ subroutine read_gmi(mype,val_gmi,ithin,rmesh,jsatid,gstime,&
 
   integer(i_kind),dimension(maxchanl)    :: tbmin                 ! different tbmin for the channels.
-  real(r_double),dimension(maxchanl)     :: mirad,gmichq,gmirfi   ! TBB from strtmbr
+!>>sieron
```
remove sieron

emilyhcliu replied:
Done!

ADCollard commented on src/gsi/read_gmi.f90:
```diff
@@ -131,7 +131,12 @@ subroutine read_gmi(mype,val_gmi,ithin,rmesh,jsatid,gstime,&
+  real(r_double),dimension(maxchanl)     :: mirad,var_check1,var_check2   ! TBB from strtmbr
+  logical                   :: use_gmichq
+!>>sieron
+
```
remove sieron

emilyhcliu replied:
Done!

ADCollard commented on src/gsi/read_gmi.f90:
```diff
@@ -131,7 +131,12 @@ subroutine read_gmi(mype,val_gmi,ithin,rmesh,jsatid,gstime,&
-  real(r_double),dimension(maxchanl)     :: mirad,gmichq,gmirfi   ! TBB from strtmbr
+!>>sieron
+  !real(r_double),dimension(maxchanl)     :: mirad,gmichq,gmirfi   ! TBB from strtmbr
+  real(r_double),dimension(maxchanl)     :: mirad,var_check1,var_check2   ! TBB from strtmbr
```
remove comment

emilyhcliu replied:
Done!

ADCollard commented on src/gsi/read_gmi.f90:
```diff
@@ -408,12 +413,25 @@ subroutine read_gmi(mype,val_gmi,ithin,rmesh,jsatid,gstime,&
-          call ufbrep(lnbufr,gmichq,1,nchanl,iret,'GMICHQ')
-          call ufbrep(lnbufr,gmirfi,1,nchanl,iret,'GMIRFI')
+!>>sieron
+          call ufbrep(lnbufr,var_check1,1,nchanl,iret,'GMICHQ')
```
remove comments

emilyhcliu replied:
Done

ADCollard commented on src/gsi/setuprad.f90:
```diff
@@ -1047,7 +1052,7 @@ subroutine setuprad(obsLL,odiagLL,lunin,mype,aivals,stats,nchanl,nreal,nobs,&
            endif
         endif
         ! Screening for cold-air outbreak area (only applied to MW for now)
```
Does this mean the cao check is only performed when precip is turned on?

emilyhcliu replied:
@ADCollard  Yes, the detection of CAO requires to use all of the five hydrometeors.

ADCollard reviewed (changes requested):
A few minor questions and changes.

I flagged a bunch of changes from Scott in TaTb which I think can be removed.

emilyhcliu commented on scripts/exglobal_atmos_analysis.sh:
```diff
@@ -81,6 +81,10 @@ export imp_physics=${imp_physics:-99}
 
+# Observation usage options
+cao_check=${cao_check:-".false."}
+ta2tb=${ta2tb:-".false."}
```
@ADCollard  First, thanks for reviewing this PR.
Yes, I will make cao_check to true as default when we do the high-resolution parallel experiment.  If the results look good, I will make cao_check=true as default in the GSI master and also in the package we hand over to NCO.

emilyhcliu commented on src/gsi/cplr_gfs_ensmod.f90:
```diff
@@ -1015,7 +1057,8 @@ subroutine parallel_read_gfsnc_state_(en_full,m_cvars2d,m_cvars3d,nlon,nlat,nsig
    deallocate(rwork3d1)
 
-   if (k3u==0.or.k3v==0.or.k3t==0.or.k3q==0.or.k3cw==0.or.k3oz==0) &
+!  if (k3u==0.or.k3v==0.or.k3t==0.or.k3q==0.or.k3cw==0.or.k3oz==0) &  !orig
```
The current one actually works for operational configuration and for the configuration with precipitation turned on without checking k3ql, k3qi, k3qr, k3qs, and k3qg.  So, this is not a strict check.

I will see if I can find a way to make it a strict check.

emilyhcliu commented on scripts/exglobal_atmos_analysis.sh:
```diff
@@ -473,6 +478,7 @@ $NLN $RTMFIX/NPOESS.VISwater.EmisCoeff.bin ./crtm_coeffs/NPOESS.VISwater.EmisCoe
 $NLN $RTMFIX/FASTEM6.MWwater.EmisCoeff.bin ./crtm_coeffs/FASTEM6.MWwater.EmisCoeff.bin
 $NLN $RTMFIX/AerosolCoeff.bin              ./crtm_coeffs/AerosolCoeff.bin
 $NLN $RTMFIX/CloudCoeff.bin                ./crtm_coeffs/CloudCoeff.bin
+#$NLN $RTMFIX/CloudCoeff.bin               ./crtm_coeffs/CloudCoeff.GFDLFV3.-109z-1.bin
```
@HaixiaLiu-NOAA Thanks for your review.
Yes, this is a typo.  I corrected the mistake as you suggested.

emilyhcliu commented on src/gsi/crtm_interface.f90:
```diff
@@ -2009,33 +2035,61 @@ subroutine call_crtm(obstype,obstime,data_s,nchanl,nreal,ich, &
+!             CRTM minimum thresholds: cloud content=1.0E-6 and cloud fraction=1.E-12
+              if (.not. regional .and. icfs==0 ) atmosphere(1)%cloud_fraction(k) = cf(kk2)
+              do ii=1,n_clouds_fwd_wk
+                 if(cloud_cont(k,ii) > 1.000_r_kind*1.0E-6_r_kind .and.  atmosphere(1)%cloud_fraction(k) < 1.001_r_kind*1.0E-12_r_kind) then
```
@HaixiaLiu-NOAA I will keep this as is.

HaixiaLiu-NOAA replied:
@emilyhcliu It would be good to keep the lines 2048 and 2089 to have consistent thresholds, I think, but I will not block this PR's approval.

emilyhcliu commented on src/gsi/prt_guess.f90:
```diff
@@ -374,7 +374,7 @@ subroutine prt_guess2(sgrep)
      nvarsc=nvarsc+1
   endif
-  if (ier/=0) return ! this is a fundamental routine, when some not found just
+! if (ier/=0) return ! this is a fundamental routine, when some not found just ! emily
```
@HaixiaLiu-NOAA I think I commented it out on purpose.  Otherwise it won't work.

emilyhcliu commented:
@MichaelLueken-NOAA 
I have the following four files change in the fix directory:
Fix files update:

 - [ ] Add global_anavinfo_allhydro_l127.txt --- add precipitation components
 - [ ] Update cloudy_radiance_info.txt --- add GMI obs error assignment and QC parameters
 - [ ] Update global_satinfo.txt --- modify channel use flag and error bounds for GMI
 - [ ] Update global_scaninfo.txt --- add GMI

They are in feature/v16x_allsky under fix directory.
What should I do to make these four fix files into rev2?    Thanks.

MichaelLueken commented:
> They are in feature/v16x_allsky under fix directory. What should I do to make these four fix files into rev2? Thanks.

@emilyhcliu You don't need to do anything.  Once the work in this PR has been merged to the master, I will need to manually merge the changes in feature/v16x_allsky into rev2.  I need to do this whether the updated fix hash is in the PR or not, so all I need is the fix branch to merge to rev2 (which you have provided) and that is it.

MichaelLueken commented on modulefiles/modulefile.ProdGSI.hera:
```diff
@@ -9,6 +9,7 @@
 #module-whatis " NOAA-EMC/GSI whatis description"
 
 setenv CRTM_FIX /scratch2/NCEPDEV/nwprod/NCEPLIBS/fix/crtm_v2.3.0
+#setenv CRTM_FIX /scratch2/NCEPDEV/nwprod/hpc-stack/libs/hpc-stack/intel-18.0.5.274/crtm/2.4.0/fix
```
Rather than add the crtm/2.4.0 fix path and module in commented out states, please either replace the crtm/2.3.0 with crtm/2.4.0 or remove the commented out lines (lines 12 and 45).

MichaelLueken commented on src/gsi/antcorr_application.f90:
```diff
@@ -153,5 +156,103 @@ SUBROUTINE Remove_AntCorr( AC  , &  ! Input
+
+    ! Arguments
+    TYPE(ACCoeff_type), INTENT(IN)     :: AC
+    INTEGER           , INTENT(IN)     :: iFOV
```
Please add:

use kinds, only: i_kind, r_kind

and add (i_kind) to integer declarations on lines 232 and 237.

MichaelLueken commented on src/gsi/antcorr_application.f90:
```diff
@@ -153,5 +156,103 @@ SUBROUTINE Remove_AntCorr( AC  , &  ! Input
+    ! Arguments
+    TYPE(ACCoeff_type), INTENT(IN)     :: AC
+    INTEGER           , INTENT(IN)     :: iFOV
+    REAL(fp)          , INTENT(IN OUT) :: T(:)
```
Please replace fp with r_kind

MichaelLueken commented on src/gsi/crtm_interface.f90:
```diff
@@ -1778,9 +1803,10 @@ subroutine call_crtm(obstype,obstime,data_s,nchanl,nreal,ich, &
 
   ! Calculate GFDL cloud fraction (if no cf in metguess table) based on PDF scheme 
-  if ( icmask .and. n_clouds_fwd_wk > 0 .and. imp_physics==11 .and.  lcalc_gfdl_cfrac ) then
+! if ( icmask .and. n_clouds_fwd_wk > 0 .and. imp_physics==11 .and.  lcalc_gfdl_cfrac ) then
```
If commented out line won't be used in the future, please remove line 1806

MichaelLueken commented on src/gsi/crtm_interface.f90:
```diff
@@ -2604,7 +2660,9 @@ subroutine calc_gfdl_cloudfrac(den,pt1,qv,cloud,hs,area,cfrac)
-  dw   = dw_ocean + (dw_land - dw_ocean) * min (one, abs(hs) / (ten * grav))
+  phi  = hs * grav   ! surface potential
+  dw   = dw_ocean + (dw_land - dw_ocean) * min (one, abs(phi) / (ten * grav))
+! dw   = dw_ocean + (dw_land - dw_ocean) * min (one, abs(hs) / (ten * grav))
```
If commented out line won't be used in the future, please remove line 2665.

MichaelLueken commented on src/gsi/general_read_gfsatm.f90:
```diff
@@ -2348,7 +2570,873 @@ subroutine general_read_gfsatm_nc(grd,sp_a,filename,uvflag,vordivflag,zflag, &
+         if ( mype == 0 ) write(6, &
+            '(a,'': different spatial dimension nlatm2 = '',i4,tr1,''latb = '',i4)') &
+            trim(my_name),nlatm2,latb
+         !call stop2(101)
```
If commented out line won't be used in the future, please removed lines 2734 and 2741.

MichaelLueken commented on src/gsi/general_read_gfsatm.f90:
```diff
@@ -2348,7 +2570,873 @@ subroutine general_read_gfsatm_nc(grd,sp_a,filename,uvflag,vordivflag,zflag, &
+   call gsi_bundlegetpointer(gfs_bundle,'ps',g_ps  ,ier);istatus=istatus+ier
+   call gsi_bundlegetpointer(gfs_bundle,'q' ,g_q   ,ier);istatus=istatus+ier
+   call gsi_bundlegetpointer(gfs_bundle,'oz',g_oz  ,ier);istatus=istatus+ier
+!  call gsi_bundlegetpointer(gfs_bundle,'cw',g_cwmr,ier);istatus=istatus+ier
```
If commented out lines won't be used in the future, please remove commented out lines 2829 and 2835.

MichaelLueken commented on src/gsi/general_read_gfsatm.f90:
```diff
@@ -2348,7 +2570,873 @@ subroutine general_read_gfsatm_nc(grd,sp_a,filename,uvflag,vordivflag,zflag, &
+      endif
+   enddo ! do k=1,nlevs
+
+!   do k=1,nlevs
+!      icount=icount+1
+!      iflag(icount)=15
+!      ilev(icount)=k
+!      kr = levs+1-k ! netcdf is top to bottom, need to flip
+!
+!      if (mype==mype_use(icount)) then
+!         call read_vardata(atmges, 'cld_amt', rwork3d0, nslice=kr, slicedim=3)
+!         ! Cloud amount (cloud fraction). 
+!         if ( diff_res ) then
+!            grid_b=rwork3d0(:,:,1)
+!            vector(1)=.false.
+!            call fill2_ns(grid_b,grid_c(:,:,1),latb+2,lonb)
+!            call g_egrid2agrid(p_high,grid_c,grid2,1,1,vector)
+!            do kk=1,grd%itotsub
+!               i=grd%ltosi_s(kk)
+!               j=grd%ltosj_s(kk)
+!               work(kk)=grid2(i,j,1)
+!            enddo
+!         else
+!            grid=rwork3d0(:,:,1)
+!            call general_fill_ns(grd,grid,work)
+!         endif
+!
+!      endif
+!      if ( icount == icm .or. k==nlevs ) then
+!         call general_reload2(grd,g_z,g_ps,g_tv,g_vor,g_div,g_u,g_v,g_q,g_oz, &
+!              g_ql,g_qi,g_qr,g_qs,g_qg,icount,iflag,ilev,work,uvflag,vordivflag,g_cf)
+!      endif
+!   enddo ! do k=1,nlevs
```
Please remove commented out code segment (lines 3348 - 3377)

MichaelLueken commented on src/gsi/general_read_gfsatm.f90:
```diff
@@ -2348,7 +2570,873 @@ subroutine general_read_gfsatm_nc(grd,sp_a,filename,uvflag,vordivflag,zflag, &
+   endif
+   deallocate(work, work_v)
+
+   ! Convert dry temperature to virtual temperature
+   !do k=1,grd%nsig
+   !   do j=1,grd%lon2
+   !      do i=1,grd%lat2
+   !         g_tv(i,j,k) = g_tv(i,j,k)*(one+fv*g_q(i,j,k))
+   !      enddo
+   !   enddo
+   !enddo
```
Please remove commented out code segment (lines 3390 - 3397)

MichaelLueken commented on src/gsi/netcdfgfs_io.f90:
```diff
@@ -155,24 +165,33 @@ subroutine read_
     type(gsi_grid)   :: atm_grid
     integer(i_kind),parameter :: n2d=2
-    integer(i_kind),parameter :: n3d=8
+    ! integer(i_kind),parameter :: n3d=8
```
If commented out line won't be used in the future, please remove it (line 168)

MichaelLueken commented on src/gsi/netcdfgfs_io.f90:
```diff
@@ -155,24 +165,33 @@ subroutine read_
+    ! integer(i_kind),parameter :: n3d=8
+    integer(i_kind),parameter :: n3d=14
     character(len=4), parameter :: vars2d(n2d) = (/ 'z   ', 'ps  ' /)
+    ! character(len=4), parameter :: vars3d(n3d) = (/ 'u   ', 'v   ', &
+    !                                                 'vor ', 'div ', &
+    !                                                 'tv  ', 'q   ', &
+    !                                                 'cw  ', 'oz  ' /)
```
Please remove commented out segment (lines 171 -174)

MichaelLueken commented on src/gsi/netcdfgfs_io.f90:
```diff
@@ -184,26 +203,61 @@ subroutine read_
        zflag=.false.
 
 !      Set values to actual MetGuess fields
+       !if (mype==0) write(*,*) 'about to call set_guess'
```
Please remove debug writes on lines 222 and 224

MichaelLueken commented on src/gsi/read_bufrtovs.f90:
```diff
@@ -506,35 +516,76 @@ subroutine read_bufrtovs(mype,val_tovs,ithin,isfcalc,&
+           end if
+
+           !write(*,*) trim(crtm_coeffs_path)
+           !write(*,*) trim(sensorlist(1))
```
Please remove debug writes on lines 537, 538, 540, and 542.

MichaelLueken commented on src/gsi/read_bufrtovs.f90:
```diff
@@ -674,22 +736,66 @@ subroutine read_bufrtovs(mype,val_tovs,ithin,isfcalc,&
+                 end if
+              else     ! EARS / DB
+                 call ufbrep(lnbufr,data1b8,1,nchanl,iret,'TMBRST')
+                 !if ( amsua .or. amsub .or. mhs .AND. sacv .ne. spc_coeff_versions)then
```
If commented out line won't be used in the future, please remove line 762.

MichaelLueken commented on src/gsi/read_bufrtovs.f90:
```diff
@@ -674,22 +736,66 @@ subroutine read_bufrtovs(mype,val_tovs,ithin,isfcalc,&
+              else     ! EARS / DB
+                 call ufbrep(lnbufr,data1b8,1,nchanl,iret,'TMBRST')
+                 !if ( amsua .or. amsub .or. mhs .AND. sacv .ne. spc_coeff_versions)then
+                 if ( amsua .or. amsub .or. mhs .AND. spc_coeff_versions .ne. 1 .AND. sacv .ne. 1)then
```
Please replace .ne. with /=

MichaelLueken commented on src/gsi/read_bufrtovs.f90:
```diff
@@ -674,22 +736,66 @@ subroutine read_bufrtovs(mype,val_tovs,ithin,isfcalc,&
+                    data1b8x=data1b8
+                    data1b4=data1b8
+                    call remove_antcorr(accoeff_sets(sacv),ifov,data1b4)
+                    !call apply_antcorr(accoeff_sets(spc_coeff_versions),ifov,data1b4)
```
Please remove commented out line 772 unless wit will be used in the future.

MichaelLueken commented on src/gsi/read_gmi.f90:
```diff
@@ -131,7 +131,10 @@ subroutine read_gmi(mype,val_gmi,ithin,rmesh,jsatid,gstime,&
 
   integer(i_kind),dimension(maxchanl)    :: tbmin                 ! different tbmin for the channels.
-  real(r_double),dimension(maxchanl)     :: mirad,gmichq,gmirfi   ! TBB from strtmbr
+  !real(r_double),dimension(maxchanl)     :: mirad,gmichq,gmirfi   ! TBB from strtmbr
```
Please remove commented out line 134 unless it will be used in the future.

MichaelLueken commented on src/gsi/read_gmi.f90:
```diff
@@ -408,12 +411,21 @@ subroutine read_gmi(mype,val_gmi,ithin,rmesh,jsatid,gstime,&
-          call ufbrep(lnbufr,gmichq,1,nchanl,iret,'GMICHQ')
-          call ufbrep(lnbufr,gmirfi,1,nchanl,iret,'GMIRFI')
+          call ufbrep(lnbufr,var_check1,1,nchanl,iret,'GMICHQ')
+          !call ufbrep(lnbufr,gmirfi,1,nchanl,iret,'GMIRFI')
```
Please remove commented out line 415 unless it will be used in the future.

MichaelLueken commented on src/gsi/read_gmi.f90:
```diff
@@ -408,12 +411,21 @@ subroutine read_gmi(mype,val_gmi,ithin,rmesh,jsatid,gstime,&
           call ufbrep(lnbufr,val_angls,n_angls,ngs,iret,'SAMA SZA SMA SGA')
           call ufbint(lnbufr,pixelloc,2, 1,iret,'CLATH CLONH')
 
+        if (any(var_check1 < 99999999999)) then   ! 100000000000 seems to be the missing value
```
Please add proper precision to real value 99999999999 (i.e. 99999999999_r_double)

MichaelLueken commented on src/gsi/read_gmi.f90:
```diff
@@ -479,13 +491,32 @@ subroutine read_gmi(mype,val_gmi,ithin,rmesh,jsatid,gstime,&
-              iskip = iskip + 1
+           if( mirad(jc)<tbmin(jc) .or. mirad(jc)>tbmax ) then
+              iskip = iskip+1
+           !endif
+           !if(use_gmichq) then
```
Please remove commented out lines 496 and 497.

MichaelLueken commented on src/gsi/read_gmi.f90:
```diff
@@ -479,13 +491,32 @@ subroutine read_gmi(mype,val_gmi,ithin,rmesh,jsatid,gstime,&
+                 nread=nread+1
+              endif
+           endif
+           !if( mirad(jc)<tbmin(jc) .or. mirad(jc)>tbmax .or. &
+           !   gmichq(jc) < -0.5_r_kind .or. gmichq(jc) > 1.5_r_kind .or. & 
+           !   gmirfi(jc)>0.0_r_kind) then ! &
+           !   iskip = iskip + 1
+           !else
+           !   nread=nread+1
+           !end if
```
If commented out segment won't be used in the future, please remove lines 513 - 519.

MichaelLueken commented on src/gsi/write_incr.f90:
```diff
@@ -235,12 +251,28 @@ subroutine write_fv3_inc_ (grd,sp_a,filename,mype_out,gfs_bundle,ibin)
+    if (iql>0) then
     call nccheck_incr(nf90_def_var(ncid_out, "liq_wat_inc", nf90_real, dimids3, liqwatvarid)) 
     call nccheck_incr(nf90_var_par_access(ncid_out, liqwatvarid, nf90_collective))
+    endif
```
Please ensure at least two space indentation for if statements.

MichaelLueken commented on src/gsi/write_incr.f90:
```diff
@@ -235,12 +251,28 @@ subroutine write_fv3_inc_ (grd,sp_a,filename,mype_out,gfs_bundle,ibin)
+    if (iqg>0) then
+    call nccheck_incr(nf90_def_var(ncid_out, "grle_inc", nf90_real, dimids3, qgvarid)) 
+    call nccheck_incr(nf90_var_par_access(ncid_out, qgvarid, nf90_collective))
+    endif
```
Please ensure at least two space indentation for if statements.

MichaelLueken commented on src/gsi/write_incr.f90:
```diff
@@ -410,6 +445,7 @@ subroutine write_fv3_inc_ (grd,sp_a,filename,mype_out,gfs_bundle,ibin)
                       start = ncstart, count = nccount))
     call mpi_barrier(mpi_comm_world,ierror)
     ! liquid water increment
+    if (iql>0) then
```
Please ensure at least two space indentation for if statement between lines 448 and 460.

MichaelLueken commented on src/gsi/write_incr.f90:
```diff
@@ -434,6 +471,7 @@ subroutine write_fv3_inc_ (grd,sp_a,filename,mype_out,gfs_bundle,ibin)
                          start = ncstart, count = nccount))
     call mpi_barrier(mpi_comm_world,ierror)
     ! ice mixing ratio increment
+    if (iqi>0) then
```
Please ensure at least two space indentation for if statement between lines 474 and 486.

MichaelLueken commented on src/gsi/write_incr.f90:
```diff
@@ -445,6 +483,49 @@ subroutine write_fv3_inc_ (grd,sp_a,filename,mype_out,gfs_bundle,ibin)
+    call nccheck_incr(nf90_put_var(ncid_out, qrvarid, sngl(out3d), &
+                      start = ncstart, count = nccount))
+    call mpi_barrier(mpi_comm_world,ierror)
+    endif
```
Please ensure at least two space indentation for if statement between lines 488 and 500.

MichaelLueken commented on src/gsi/write_incr.f90:
```diff
@@ -445,6 +483,49 @@ subroutine write_fv3_inc_ (grd,sp_a,filename,mype_out,gfs_bundle,ibin)
+    call nccheck_incr(nf90_put_var(ncid_out, qsvarid, sngl(out3d), &
+                      start = ncstart, count = nccount))
+    call mpi_barrier(mpi_comm_world,ierror)
+    endif
```
Please ensure at least two space indentation for if statement between lines 502 and 514.

MichaelLueken commented on src/gsi/write_incr.f90:
```diff
@@ -445,6 +483,49 @@ subroutine write_fv3_inc_ (grd,sp_a,filename,mype_out,gfs_bundle,ibin)
+    call nccheck_incr(nf90_put_var(ncid_out, qgvarid, sngl(out3d), &
+                      start = ncstart, count = nccount))
+    call mpi_barrier(mpi_comm_world,ierror)
+    endif
```
Please ensure at least two space indentation for if statement between lines 516 and 528.

MichaelLueken reviewed (commented):
Hi @emilyhcliu, I have completed my initial review of your changes.  The first issue is that there are four commit messages.  To reduce this to a single commit message, please do the following in your feature/v16x_allsky branch in your fork:

1) git reset --soft HEAD~1
2) git reset --soft HEAD~1
3) git reset --soft HEAD~1
4) git commit --amend

In the commit message window, please add:

`GitHub Issue NOAA-EMC/GSI#341`

to the commit message:

```
GitHub Issue NOAA_EMC/GSI#341. This update includes the following:
(1) Add the capability of assimilating precipitation-affected radiances from AMSU-A and ATMS
(2) Enable the all-sky radiance assimilation for GMI
(3) Add the switch to use brightness temperature (SDR) instead of antenna temperature (TDR) for AMSU-A, ATMS and MHS
```
Once the commit messages have been corrected, please go through and address the concerns I noted (mostly, removal of commented out lines of code that if they won't be used in the future).  Once the changes have been made, please use:

1) git commit --amend
2) Save and close the commit message without changing the message.
3) Push back to the repository using
     `git push origin feature/v16x_allsky --force`

Please let me know if you have any questions, comments, or concerns.

MichaelLueken commented on src/gsi/general_read_gfsatm.f90:
```diff
@@ -2348,7 +2570,873 @@ subroutine general_read_gfsatm_nc(grd,sp_a,filename,uvflag,vordivflag,zflag, &
+   real(r_kind),pointer,dimension(:,:)       :: g_ps
+   real(r_kind),pointer,dimension(:,:,:)     :: g_vor,g_div,&
+                                                g_q,g_oz,g_tv
+   real(r_kind),pointer,dimension(:,:,:)     :: g_ql,g_qi,g_qr,g_qs,g_qg,g_cf
```
While compiling in debug mode, g_cf, on line 2644, was noted as being unused.  All code segments that use this variable are commented out.  If these sections won't be used in the future, please remove g_cf.  Otherwise, please comment out g_cf until the other commented out sections are uncommented.

MichaelLueken commented on src/gsi/general_read_gfsatm.f90:
```diff
@@ -2348,7 +2570,873 @@ subroutine general_read_gfsatm_nc(grd,sp_a,filename,uvflag,vordivflag,zflag, &
 
 end subroutine general_read_gfsatm_nc
+subroutine general_read_gfsatm_allhydro_nc(grd,sp_a,filename,uvflag,vordivflag,zflag, &
+           gfs_bundle,init_head,iret_read)
```
On line 2574, init_head was identified as unused while compiling in debug mode.  Please remove unused variable.

MichaelLueken commented on src/gsi/netcdfgfs_io.f90:
```diff
@@ -128,12 +130,16 @@ subroutine read_
     character(24) filename
     integer(i_kind):: it, istatus, inner_vars, num_fields
-    integer(i_kind):: iret_ql,iret_qi
+    integer(i_kind):: iret_ql,iret_qi,iret_qr,iret_qs,iret_qg
```
On line 140, iret_ql, iret_qi, iret_qr, iret_qs, and iret_qg were all identified as unused while compiling in debug mode.  Please remove unused variables.

MichaelLueken commented on src/gsi/write_incr.f90:
```diff
@@ -91,6 +91,9 @@ subroutine write_fv3_inc_ (grd,sp_a,filename,mype_out,gfs_bundle,ibin)
     use obsmod, only: ianldate 
     use state_vectors, only: allocate_state, deallocate_state
 
+    use state_vectors, only: svars3d, ns3d
```
On line 94, ns3d was noted as unused.  Please remove unused variable.

MichaelLueken commented on src/gsi/write_incr.f90:
```diff
@@ -128,7 +132,9 @@ subroutine write_fv3_inc_ (grd,sp_a,filename,mype_out,gfs_bundle,ibin)
-                       tvarid, sphumvarid, liqwatvarid, o3varid, icvarid
+                       tvarid, sphumvarid, liqwatvarid, o3varid, icvarid, &
+                       qrvarid, qsvarid, qgvarid
+    integer(i_kind) :: icw,iql,iqi,iqr,iqs,iqg
```
On line 137, icw was identified as unused while compiling in debug mode.  Please remove unused variable.

MichaelLueken commented on src/gsi/qcmod.f90:
```diff
@@ -3132,12 +3130,34 @@ subroutine qc_amsua(nchanl,is,ndat,nsig,npred,sea,land,ice,snow,mixed,luse,   &
+                    end if
                  end if
+                ! test in the future
+                ! if (si_mean >= 20.0_r_kind) then
```
Since si_mean is currently commented out, si_mean on line 2810 is now being flagged as unused.  If possible, please move si_mean to the end of the passed variable list and comment it out until this section is brought back.  Please note that this will also require some additional work in setuprad.f90.

emilyhcliu replied:
@MichaelLueken-NOAA I made si_mean an optional input.

MichaelLueken reviewed (commented):
Hi @emilyhcliu, I have completed my debug review of your changes in this PR.  There are several unused variables that were flagged while compiling and I have noted them in this review.  Please remove or comment out the unused variables (I say comment out because several of the unused variables are being flagged as such due to the code segments that they are used being commented out).  Having said that, the debug executable ran through to completion, passing the debug tests.  If you have any questions or concerns, please let me know.

emilyhcliu commented:
@MichaelLueken-NOAA I fixed the commit comment as you instructed.  I also addressed your comments and changed according.  I had committed and push the changes back to the branch.  Please let me know if you have any questions/comments.

MichaelLueken commented:
@emilyhcliu Your PR has too many commit messages.  You will need to reduce the six commits down to one using:

1) git reset --soft HEAD~1
2) git commit --amend
3) Please save and close the commit message window without changing the commit message
4) git push origin feature/v16x_allsky --force

If you have any questions, please let me know.

emilyhcliu commented:
> 

Done!

MichaelLueken commented:
Hi @emilyhcliu.  Thank you very much for addressing my concerns.  There are just a few things I would like to note quickly:

1) Unused variables:
    * Following an updated clone of your fork, I see that there are still two unused variables:
       * src/gsi/general_read_gfsatm.f90 - g_cf on line 2642.  Please comment out unused variable.
       * src/gsi/qcmod.f90 - si_mean on lines 2815 and 3623.  Please comment out unused variables (changing passed variables to optional doesn't remove the fact that they aren't used in the code).
2) Commented out code:
    * If the currently commented out code that hasn't been removed will be used in the future, then it can stay.  Please note that it is not desirable to keep commented out code in routines, so if there is no chance that the commented out code will be used again, please remove it.  However, it looks like there are commented out lines that are purely for debug purposes:
       * src/gsi/netcdfgfs_io.f90 - please remove debug writes on lines 220 and 222.
       * src/gsi/read_bufrtovs.f90 - please remove debug writes on lines 537, 538, 540, and 542.
3) Indentation:
    * src/gsi/write_incr.f90 - Please ensure two space indentation for all if statements between lines 254 - 275.
4) Real precision:
    * src/gsi/read_gmi.f90 - The variable, var_check1, is declared as r_double.  On line 420, the value 99999999999 isn't given a precision.  Please ensure that the proper precision is given to this value (whether it is _r_kind or _r_double).
5) Logical Boolean symbols:
    * src/gsi/read_bufrtovs.f90 - Please replace .ne. with /= on line 763.

After changes are made, please commit using:

git commit --amend

and push using:

git push origin feature/v16x_allsky --force

If you have any questions, please let me know.

MichaelLueken commented on src/gsi/crtm_interface.f90:
```diff
@@ -2009,33 +2035,61 @@ subroutine call_crtm(obstype,obstime,data_s,nchanl,nreal,ich, &
+                     cloud_efr(k,ii)=max(5.001_r_kind, cloud_efr(k,ii))
+                 endif
+              end do
+!crtm2.3.x    if (.not. regional .and. icfs==0 ) atmosphere(1)%cloud_fraction(k) = cf(kk2)
```
Should line 2083 be removed, since an uncommented out version of the line exists on line 2087?

emilyhcliu replied:
@MichaelLueken-NOAA You were right.  I removed them.

MichaelLueken commented on src/gsi/crtm_interface.f90:
```diff
@@ -2000,7 +2026,7 @@ subroutine call_crtm(obstype,obstime,data_s,nchanl,nreal,ich, &
                     cloud_cont(k,2)=max(1.001_r_kind*1.0E-6_r_kind, cloud_cont(k,2))
               end do
-!crtm2.3.x    if (.not. regional .and. icfs==0 ) atmosphere(1)%cloud_fraction(k) = cf(kk2)   
+!crtm2.3.x    if (.not. regional .and. icfs==0 ) atmosphere(1)%cloud_fraction(k) = cf(kk2)
```
Should line 2029 be removed, since an uncommented out version of the line exists on line 2087?

MichaelLueken reviewed (commented):
Hi @emilyhcliu.  Two more minor points in src/gsi/crtm_interface.f90.  There are several commented out lines pertaining to CRTM v2.3.* (specifically, lines 2029 and 2083).  Both of these lines appear to be identical.  Additionally, they are identical to line 2087, which is not commented.  Should these two commend out lines be removed?

emilyhcliu commented:
@MichaelLueken-NOAA I updated the codes as you suggested and pushed the back.  
Thanks for reviewing. Please let me know if you have any comments.

MichaelLueken commented on src/gsi/read_bufrtovs.f90:
```diff
@@ -674,22 +732,66 @@ subroutine read_bufrtovs(mype,val_tovs,ithin,isfcalc,&
+              else     ! EARS / DB
+                 call ufbrep(lnbufr,data1b8,1,nchanl,iret,'TMBRST')
+                 !if ( amsua .or. amsub .or. mhs .AND. sacv .ne. spc_coeff_versions)then
+                 if ( amsua .or. amsub .or. mhs .AND. spc_coeff_versions /= 1 .AND. sacv .ne. 1)then
```
Thank you for replacing the first .ne. with /=.  Please replace the second .ne. with /= on line 759.

emilyhcliu replied:
@MichaelLueken-NOAA Thanks you for your thorough review of my PR.  Your request has been addressed.

MichaelLueken reviewed (commented):
Hi @emilyhcliu.  Thank you very much for making the requested changes.  There is one last minor change.  On line 759 of src/gsi/read_bufrtovs.f90, you replaced the first .ne. with /=, as requested.  However, there is a second .ne. (sacv .ne. 1) that needs to also be replaced with /= (i.e., sacv /= 1).  Once this change is made, I should be able to submit this work to the review committee once the current work has been merged to the authoritative repository.

After making the change, please use:

1) git add src/gsi/read_bufrtovs.f90
2) git commit --amend
3) Save and close the commit message window without altering the current commit message
4) git push origin feature/v16x_allsky --force

If you have any questions, please let me know.

MichaelLueken commented:
The due date for the review committee has passed without feedback.  I will now give final approval to these changes and merge them to the authoritative repository.

MichaelLueken commented:
The fix files in fix/feature/v16x_allsky were merged to fix/rev2 and pushed to the authoritative repository at [87fe77d](https://github.com/NOAA-EMC/GSI/commit/87fe77d7f2d456252235a9bc302c62fff6ea7fe9).
